pkg/git/remote_git_resolver.go

```diff
@@ -48,16 +77,66 @@ func (*remoteGitResolver) Resolve(auth transport.AuthMethod, sourceConfig corev1
 			}, nil
 		}
 	}
+	return corev1alpha1.ResolvedSourceConfig{}, err
+}
+
+func resolveSourceWithSubpath(auth transport.AuthMethod, sourceConfig corev1alpha1.SourceConfig) (corev1alpha1.ResolvedSourceConfig, error) {
+	var resolvedConfig corev1alpha1.ResolvedSourceConfig
+
+	r, err := gogit.Clone(memory.NewStorage(), nil, &gogit.CloneOptions{
+		URL:  sourceConfig.Git.URL,
+		Auth: auth,
+	})
+
+	rIter, _ := r.References()
+
+	rIter.ForEach(func(ref *plumbing.Reference) error {
+		if ref.Name().Short() == sourceConfig.Git.Revision {
+			refSourceType := sourceType(ref)
+			effectiveCommit := ""
+
+			if refSourceType == corev1alpha1.Branch {
+				effectiveCommit, err = latestCommitForSubpath(r, sourceConfig.SubPath)
+			} else {
+				effectiveCommit = ref.Hash().String()
+			}
+
+			if err != nil {
+				return nil
+			}
 
-	return corev1alpha1.ResolvedSourceConfig{
-		Git: &corev1alpha1.ResolvedGitSource{
-			URL:                  sourceConfig.Git.URL,
-			Revision:             sourceConfig.Git.Revision,
-			Type:                 corev1alpha1.Commit,
-			SubPath:              sourceConfig.SubPath,
-			InitializeSubmodules: sourceConfig.Git.InitializeSubmodules,
+			resolvedConfig = corev1alpha1.ResolvedSourceConfig{
+				Git: &corev1alpha1.ResolvedGitSource{
+					URL:                  sourceConfig.Git.URL,
+					Revision:             effectiveCommit,
+					Type:                 refSourceType,
+					SubPath:              sourceConfig.SubPath,
+					InitializeSubmodules: sourceConfig.Git.InitializeSubmodules,
+				},
+			}
+		}
+		return nil
+	})
+
+	return resolvedConfig, err
+}
+
+func latestCommitForSubpath(r *gogit.Repository, subPath string) (string, error) {
+	logOutput, err := r.Log(&gogit.LogOptions{
+		PathFilter: func(s string) bool {
+			if strings.HasPrefix(s, subPath) {
+				return true
+			} else {
+				return false
+			}
 		},
-	}, nil
+	})
+
+	latestCommit, err := logOutput.Next()
+	regex := regexp.MustCompile(commitRegex)
```

Comment:
Let's move this regex to a top level var, compiling a regex is non-trivial and doing it every minute per git source for the exact same static regex is not ideal.